driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -184,92 +184,95 @@
+            int colonCount = countOccurrences(userInfo, ":");
+            if (userInfo.contains("@") || colonCount > 1) {
+                throw new IllegalArgumentException("The connection string contains invalid user information. "
+                        + "If the username or password contains a colon (:) or an at-sign (@) then it must be urlencoded");
             }
+            if (colonCount == 0) {
+                userName = urldecode(userInfo);
+            } else {
+                idx = userInfo.indexOf(":");
+                userName = urldecode(userInfo.substring(0, idx));
+                password = urldecode(userInfo.substring(idx + 1), true).toCharArray();
+            }
+        } else {
+            hostIdentifier = userAndHostInformation;
+        }
 
-            Collections.addAll(all, serverPart.split(","));
-
-            Collections.sort(all);
-            hosts = Collections.unmodifiableList(all);
+        // Validate the hosts
+        hosts = Collections.unmodifiableList(parseHosts(asList(hostIdentifier.split(","))));
 
-            if (nsPart != null && !nsPart.isEmpty()) { // database,_collection
-                idx = nsPart.indexOf(".");
-                if (idx < 0) {
-                    database = nsPart;
-                    collection = null;
-                } else {
-                    database = nsPart.substring(0, idx);
-                    collection = nsPart.substring(idx + 1);
-                }
-            } else {
-                database = null;
+        // Process the authDB section
+        String nsPart = null;
+        idx = unProcessedConnectionString.indexOf("?");
+        if (idx == -1) {
+            nsPart = unProcessedConnectionString;
+            unProcessedConnectionString = "";
+        } else {
+            nsPart = unProcessedConnectionString.substring(0, idx);
+            unProcessedConnectionString = unProcessedConnectionString.substring(idx + 1);
+        }
+        if (nsPart.length() > 0) {
+            nsPart = urldecode(nsPart);
+            idx = nsPart.indexOf(".");
+            if (idx < 0) {
+                database = nsPart;
                 collection = null;
+            } else {
+                database = nsPart.substring(0, idx);
+                collection = nsPart.substring(idx + 1);
             }
+        } else {
+            database = null;
+            collection = null;
+        }
 
-            Map<String, List<String>> optionsMap = parseOptions(optionsPart);
-            translateOptions(optionsMap);
+        Map<String, List<String>> optionsMap = parseOptions(unProcessedConnectionString);
+        translateOptions(optionsMap);
+        try {
             credentials = createCredentials(optionsMap, userName, password);
-            warnOnUnsupportedOptions(optionsMap);
-        } catch (UnsupportedEncodingException e) {
-            throw new MongoInternalException("This should not happen", e);
+        } catch (Throwable t) {
```

Comment:
Generally a bad practice to catch Throwable.  Better to catch RuntimeException and let Error subclasses through.  But why catch and re-throw at all rather than just throw whatever createCredentials throws? (I tried removing this catch and all the tests still pass).